packages/fast-components/src/dialog/dialog.ts

```diff
@@ -0,0 +1,145 @@
+import { attr, FastElement } from "@microsoft/fast-element";
+import { keyCodeEscape, keyCodeTab } from "@microsoft/fast-web-utilities";
+import tabbable from "tabbable";
+
+export class Dialog extends FastElement {
+    @attr
+    public modal: boolean = true;
+
+    @attr
+    public hidden: boolean = false;
+
+    @attr({ attribute: "trap-focus" })
+    public trapFocus: boolean = true;
+
+    @attr({ attribute: "aria-describedby" })
+    public ariaDescribedby: string;
+
+    @attr({ attribute: "aria-labelledby" })
+    public ariaLabelledby: string;
+
+    @attr({ attribute: "aria-label" })
+    public ariaLabel: string;
+
+    public dialog: HTMLDivElement;
+
+    public dismiss(): void {
+        this.dispatchEvent(
+            new CustomEvent("dismiss", {
+                bubbles: true,
+                composed: true,
+            })
+        );
+    }
+
+    public connectedCallback(): void {
+        super.connectedCallback();
+        document.addEventListener("keydown", this.handleDocumentKeydown);
+
+        if (this.shouldDialogTrapFocus()) {
+            document.addEventListener("focusin", this.handleDocumentFocus);
+
+            if (this.shouldForceFocus(document.activeElement)) {
+                this.focusFirstElement();
+            }
+        }
+    }
+
+    public disconnectedCallback(): void {
+        super.disconnectedCallback();
+
+        document.removeEventListener("keydown", this.handleDocumentKeydown);
+
+        if (this.shouldDialogTrapFocus()) {
+            document.removeEventListener("focusin", this.handleDocumentFocus);
+        }
+    }
+
+    private handleDocumentKeydown = (e: KeyboardEvent): void => {
+        // remove as any here when we have boolean support
+        if (!e.defaultPrevented && !this.isDialogHidden()) {
+            switch (e.keyCode) {
+                case keyCodeEscape:
+                    this.dismiss();
+                    break;
+
+                case keyCodeTab:
+                    this.handleTabKeyDown(e);
+                    break;
+            }
+        }
+    };
+
+    private handleDocumentFocus = (e: Event): void => {
+        if (!e.defaultPrevented && this.shouldForceFocus(e.target as HTMLElement)) {
+            this.focusFirstElement();
+            e.preventDefault();
+        }
+    };
+
+    private handleTabKeyDown = (e: KeyboardEvent): void => {
+        if (!this.shouldDialogTrapFocus()) {
+            return;
+        }
+
+        const tabbableElements: HTMLElement[] = tabbable(this);
```

Comment:
Invoking `tabbable` every keyDown is going to end up querying the DOM a lot more than is likely necessary. Perhaps we should store the tabbable element references and then use a mutation observer to update the list when elements are added / removed?

Reply:
Yeah I was thinking about that but decided to push this out and get feedback (like this).

Reply:
taken another go at this based on feedback. Resolving this only as it's outdated.